### PR TITLE
Store tape for gradient

### PR DESCRIPTION
**Pull request by tmigot**

### CODE REVIEW

**Comment by tmigot** [comment]
'NLPModelsIpopt.jl' fails because it used explicitly 'ADNLPModels.ForwardDiffAD(::Int64)'.

**Comment by tmigot** [comment]
As a small benchmark of this PR:
```
using Pkg
#Pkg.add(url="https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl#main")
#Pkg.add(url="https://github.com/tmigot/ADNLPModels.jl#tape-gradient")
branch = "tape-gradient" # "main"
Pkg.activate(branch)
#This package
using BenchmarkTools, ForwardDiff, NLPModels, ReverseDiff
backend = "ADNLPModels.ReverseDiff" # "ADNLPModels.ForwardDiff"
using ADNLPModels

n = 100000
function f(x)
    n = length(x)
    return 1 +
           100 * sum((x[i + 1] - x[i]^2)^2 for i = 1:(n - 1)) +
           sum((x[i] - 1)^2 for i = 1:(n - 1))
end
x0 = [i / (n + 1) for i = 1:n];

if branch == "main"
    nlp = eval(Meta.parse("ADNLPModel(f, x0, adbackend=" * backend * "AD($n))"))
else
    nlp = eval(Meta.parse("ADNLPModel(f, x0, adbackend=" * backend * "AD($n, $f, $x0))"))
end
g = similar(nlp.meta.x0)
x = nlp.meta.x0;
@btime grad!(nlp, x, g);
x = rand(nlp.meta.nvar);
@btime grad!(nlp, x, g);
```

ForwardDiff - main
  5.507 s (3 allocations: 9.92 MiB)
  5.410 s (3 allocations: 9.92 MiB)

ForwardDiff - tape-gradient
  5.228 s (1 allocation: 1.14 KiB)
  5.192 s (1 allocation: 1.14 KiB)

ReverseDiff - main
  135.298 ms (4099990 allocations: 161.59 MiB)
  137.226 ms (4099990 allocations: 161.59 MiB)

ReverseDiff - tape-gradient
  30.344 ms (0 allocations: 0 bytes)
  30.421 ms (0 allocations: 0 bytes)

File: docs/make.jl [hunk 1]
```diff
@@ -13,4 +13,8 @@ makedocs(
   pages = ["Home" => "index.md", "Tutorial" => "tutorial.md", "Reference" => "reference.md"],
 )
 
-deploydocs(repo = "github.com/JuliaSmoothOptimizers/ADNLPModels.jl.git", push_preview = true, devbranch = "main")
+deploydocs(
+  repo = "github.com/JuliaSmoothOptimizers/ADNLPModels.jl.git",
+  push_preview = true,
+  devbranch = "main",
+)
```
Comment (dpo) [inline docs/make.jl:20]: This seems unrelated to this PR?!

**Review by dpo — COMMENTED** [review]
just a minor comment

File: docs/make.jl (hunk 1 above)
Comment (tmigot) [reply docs/make.jl]: sorry for that, it was a "squashing" issue.

**Review by dpo — APPROVED** [review]

**Comment by dpo** [comment]
Thanks!
